modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -89,86 +157,468 @@ public IgniteStatisticsRepository statisticsRepository() {
     }
 
     /** {@inheritDoc} */
-    @Override public void clearObjectStatistics(String schemaName, String objName, String... colNames) {
-        StatsKey key = new StatsKey(schemaName, objName);
+    @Override public void clearObjectStatistics(
+            String schemaName,
+            String objName,
+            String... colNames
+    ) throws IgniteCheckedException {
+        final StatsCollectionFutureAdapter doneFut = new StatsCollectionFutureAdapter();
+
+        UUID reqId = UUID.randomUUID();
+        StatsKeyMessage keyMsg = StatisticsUtils.toMessage(schemaName, objName, colNames);
+
+        Map<UUID, List<Integer>> requestNodes = null;
+        /*try {
+            requestNodes = nodePartitions(extractGroups(Collections.singletonList(keyMsg)), null);
+        } catch (IgniteCheckedException e) {
+            // TODO: handle & remove task
+        }
+
+        StatsClearRequest req = new StatsClearRequest(reqId, false, Collections.singletonList(keyMsg));
+
+        sendLocalRequests(reqId, req, requestNodes);*/
+
+        clearObjectStatisticsLocal(keyMsg);
+    }
+
+    /**
+     * Actually clear local object statistics by the given key.
+     *
+     * @param keyMsg Key to clear statistics by.
+     */
+    private void clearObjectStatisticsLocal(StatsKeyMessage keyMsg) {
+        StatsKey key = new StatsKey(keyMsg.schema(), keyMsg.obj());
+        String[] colNames = keyMsg.colNames().toArray(new String[0]);
+
         statsRepos.clearLocalPartitionsStatistics(key, colNames);
         statsRepos.clearLocalStatistics(key, colNames);
         statsRepos.clearGlobalStatistics(key, colNames);
     }
 
+    // TODO
+    /**
+     * Collect local object statistics by primary partitions of specified object.
+     *
+     * @param keyMsg Statistic key message to collect statistics by.
+     * @param partIds Set of partition ids to collect statistics by.
+     * @param cancelled Supplier to check if operation was cancelled.
+     * @return Tuple of Collected local statistics with array of successfully collected partitions.
+     * @throws IgniteCheckedException
+     */
+    private IgniteBiTuple<ObjectStatisticsImpl, int[]> collectLocalObjectStatistics(
+            StatsKeyMessage keyMsg,
+            int[] partIds,
+            Supplier<Boolean> cancelled
+    ) throws IgniteCheckedException {
+        GridH2Table tbl = schemaMgr.dataTable(keyMsg.schema(), keyMsg.obj());
+        if (tbl == null)
+            throw new IgniteCheckedException(String.format("Can't find table %s.%s", keyMsg.schema(), keyMsg.obj()));
+
+        Column[] selectedCols = filterColumns(tbl.getColumns(), keyMsg.colNames());
+
+        Collection<ObjectPartitionStatisticsImpl> partsStats = collectPartitionStatistics(tbl, partIds, selectedCols,
+                cancelled);
```

Comment:
Why do we need to pass a Cancel supplier as it look like method is executed synchronously?

Reply:
Can statistic be calculated for partitions in parallel?

Reply:
We pass cancelled to allow to cancel such collection while it works. It'll happen when CancelStatisticsCollectionRequest received. Here cancelled - is just wrapper around statistics collection status doneFuture. So to cancel collection we'll just get status from current collection map and mark it's future as cancelled (and remove status from current map so worker won't event find it there and stop request processing).

Reply:
For now - no, all collection made by single thread from IgniteStatisticsManagerImpl.statMgmtPool. Only cancel and reschedule operation can be conducted in parallel (but not for single object during synchronization). But in general - it should be possible to collect statistics in parallel.